Added possibility to retrieve the raw socket and loop externally

thg2k opened the pull request:
When multiplexing between different streams there is the need to use stream_select() on the raw sockets, and then run for a single time into the main loop of the library. These two changes allow to do so.

Namoshek commented:
Thanks for the suggestion! Can you elaborate a bit on the use case for this and/or give a code snippet where you'd use it?

I'm currently in the process of refactoring the client in order to support MQTT 3.1.1 and 5.0 in the future. As part of this refactoring, I was considering to make use of [reactphp](https://github.com/reactphp/reactphp) or [amp](https://github.com/amphp/amp) for the event loop. Would such a change be helpful for your use case?

thg2k commented:
> Thanks for the suggestion! Can you elaborate a bit on the use case for this and/or give a code snippet where you'd use it?
> 

Well, I'm not a big fan of frameworks, I always tend to write my own code and pull in packages like yours for very specific tasks like the MQTT communication, which is a simple and straightforward protocol. That is to say, I wrote my own main loop with a stream polling subroutine, but I had to make this change to make sure that:

1) You don't hijack the flow, I do my own stream_select() cause I await data from another TCP socket and I need to keep the control of the blocking points to ensure quick reactions
2) Because of the above, I needed access to the raw stream to know when there is incoming data from subscriptions, in which case I simply call your loop() but I want the data processed and the control returned right away. I also implemented a short loop timeout (5 seconds) so I give control to the library to check for timeouts in case it's time to ping the server.

These two changes are enough for integrating your library, I still need to do more testing but so far it seems to work like a charm, I just need a way to escape your main loop *even* if there are subscriptions, otherwise I can't read data from the TCP socket as soon as it's available. Do you have better ideas to solve my problem?

(OT: I'll send another pull request this week about the TLS implementation)
(PS: I'm happy to take a look at your refactoring if you push it to the repo as WIP).

thg2k commented:
Below is most of my façade implementation for your library, I also forgot to mention that I need to keep in touch with multiple MQTT servers, so I instantiate your library more than once. In particular, I collect data from a TCP socket, manipulate it and broadcast to several MQTT servers.

```php

use PhpMqtt\Client\MQTTClient;
use PhpMqtt\Client\Logger as MQTTLogger;

class _mqtt_client_logger
        extends MQTTLogger
{
  private static $_instance_seq = 0;

  private $_instance_id;

  public function __construct()
  {
    self::$_instance_seq++;
    $this->_instance_id = self::$_instance_seq;
  }

  public function log($level, $message, array $context = []): void
  {
    xdbg(10, "[mqtt] CLIENT(" . $this->_instance_id . "): LOG($level): $message" .
        (count($context) ? "\n" . var_export($context, true) : ""));
  }
}

/**
 * ...
 */
class mqtt_client
        implements stream_pollable
{
  private $_client;

  private $_config;

  public function __construct(array $config)
  {
    $this->_config = array();

    $this->_config['host'] = ($config['server'] ?? null);

    $this->_config['port'] = ($config['port'] ?? 1883);

    $this->_config['client_id'] = ($config['client_id'] ?? "");

    $this->_config['username'] = ($config['username'] ?? "");

    $this->_config['password'] = ($config['password'] ?? "");

    $this->_config['tls_enabled'] = ($config['tls_enabled'] ?? false);
    if (!is_bool($this->_config['tls_enabled']))
      throw new \InvalidArgumentException(
          "Invalid config entry 'tls_enabled': must be bool");

    $this->_client = new MQTTClient(
        $this->_config['host'],
        $this->_config['port'],
        ($this->_config['client_id'] ?: null),
        null,
        null,
        new _mqtt_client_logger());

    $this->_client->connect(
        ($this->_config['username'] ?: null),
        ($this->_config['password'] ?: null));

    /* initialize the wrapped client */
  }



  public function stream_poll_register()
  {
    xdbg(30, "[mqtt] registering polling fds");

    // $_client_socket = $this->_client->socket;

    $_client_socket = $this->_client->getSocket();

    // FIXME: extract socket descriptor from client library
    // $robj = new \ReflectionObject($this->_client);
    // $rfield = $robj->getProperty('socket');
    // $rfield->setAccessible(true);
    // $_client_socket = $rfield->getValue($this->_client);

    return array(array($_client_socket), null, null, 5);
  }

  public function stream_poll_event($type, $fd)
  {
    xdbg(10, "[mqtt] poll event of type '$type'");

    if ($type == 0) { // FIXME: constants?
      if ($fd) {
        $this->_client->loop(false, true, -1);
      }
      else
        throw new \RuntimeException("Bogus fd as stream_poll_event() [read]");
    }
    elseif ($type == -1) {
      $this->_client->loop(false, true, -1);
    }
    else
      throw new \RuntimeException("Bogus event type '$type' as stream_poll_event()");
  }

  public function publish($topic, $data)
  {
    xdbg(30, "[mqtt] (on " . $this->getUrl() . ") Publishing topic '$topic' data '$data'");
    $this->_client->publish($topic, $data);
  }

  public function subscribe($topic, $callback, $qos = 0)
  {
    xdbg(30, "[mqtt] (on " . $this->getUrl() . ") Subscribing to '$topic'");
    $this->_client->subscribe($topic, $callback, $qos);
  }
}

```

Namoshek commented:
Interesting use case! I need to think this through, maybe there is a better solution using threads or something like that.

Regarding your suggestion, I have two comments:
- I'm ok with adding the `socket()` method to the `MQTTClient` class, but I'd like to keep it off the contract (interface) for now, since I don't want to guarantee it to stay. Allowing others to fiddle with the internal socket may just result in a painful experience after all.

- Exiting the loop should be possible with the current options already. You can register a `$loopEventHandler` and interrupt the loop:
  ```php
  $mqtt = new MQTTClient($host, $port);
  $mqtt->registerLoopEventHandler(function ($client) {
      $client->interrupt();
  });
  $mqtt->loop(); // runs once only (exits when starting 2nd loop)
  ```
  In the `refactoring` branch, I already added an example in the description of the hooks: [src/Concerns/OffersHooks.php](https://github.com/php-mqtt/client/blob/refactoring/src/Concerns/OffersHooks.php)
  Calling `interrupt()` should be safe in your case, since the `$interrupt` signal is reset immediately when the loop follows it.

thg2k commented:
Ahh, I totally missed the `interrupt()` method. I'm deeply sorry. In this case, I have everything I need, I can force the socket out using reflection, but of course it's still nicer to have it in the interface.

I took a look at ReactPHP and amphp, they look really cool but feels like a lot of code, I'm working on an embedded system so i don't feel like pulling in a lot of boilerplate code. What we would need is a PSR standard for event loops so I can make my own one and still use a library like yours using only the minimal functionality that I need.

Namoshek commented:
Sorry for the late answer. That sounds great!

Maybe I will provide the raw socket as well as the "inner part" of the loop as `protected` methods in the upcoming refactoring release. I will then make clear that these are availably options for people who know what they are doing, and that these options are only available if you inherit your own client from the provided one. It will not be necessary to work with reflection then anymore.

I'll leave this issue open for now and notify you, when I'm done refactoring, so you can have a look.

thg2k commented:
*OFFTOPIC*
Sorry for the offtopic, I don't know how to contact you directly. Do you have an email address where I can write to?
Anyway: I am hitting various shortcomings with your library, would you be available for a chat/call to discuss these? I may be interested to hop on your refactoring effort and help out if we have a similar vision on how it should be.

Some examples:
1. TCP disconnections are not handled well, I would expect the client to automatically reconnect with exponential backoff
2. Messages are not stored in the memory repository during disconnect time, I would be happy to have
3. TLS connection options are not flexible enough: I want to be able to choose:
  a) whether I am willing to take unverified peer certificate (at my own risk)
  b) start tls using system bundled CAs for verification

I have more smaller issues that would require some major refactoring and moving code around, so let me know if you are interested, otherwise I'll just patch what I need and move on.

Thank you a lot

Namoshek commented:
I'm always open for support and contributions, and I can see where you are coming from with your suggestions. You can reach me in Discord under the name Namoshek#9741 or via email (checkout the git history, I'm not to keen posting one on indexed sites 👍).

thg2k commented:
Hello! I rebased and updated the pull request. You were right, having the `$socket` property protected is a wise choice, it still allows you to extract it by extending the class, but suggests you are going to look for troubles. Much better than a public getter method.

I still kept the hack with waiting time -1 as I'm not happy with having the whole callback situation invoking `interrupt()`. If the interrupted flag was checked at the end of the loop instead of the beginning I could just interrupt it just before calling `loop()`, but that might have behavioural changes hard to predict.

It is up to you whether to merge this or not, it is a stable branch afterall and I can live with my own patched fork until the shiny new refactoring branch becomes a reality :smile:

Namoshek commented on src/MQTTClient.php:
```diff
@@ -700,6 +700,12 @@ public function loop(bool $allowSleep = true, bool $exitWhenQueuesEmpty = false,
                     break;
                 }
             }
+
+            // Exit for real the loop after one round if limit -1
+            if ($queueWaitLimit == -1) {
+                $this->logger->debug('Force-exiting the main loop after one go');
+                break;
+            }
```
I don't quite like this yet because I find it very non-intuitive. Can we use a different solution adding a new parameter (with a default)? One of these two solutions looks good to me:

*1) `$maxIterations` or `$exitAfterIterations`*
```php
 public function loop(
    bool $allowSleep = true, 
    bool $exitWhenQueuesEmpty = false, 
    int $queueWaitLimit = null,
    int $maxIterations = null
): void
{
    $iteration = 1;

    while (true) {
        // Do normal work

        if ($maxIterations !== null && $maxIterations <= $iteration) {
            break;
        }

        $iteration++;
    }
}
```

*2) `$runOnlyOnce`*
```php
 public function loop(
    bool $allowSleep = true, 
    bool $exitWhenQueuesEmpty = false, 
    int $queueWaitLimit = null,
    int $runOnlyOnce = false
): void
{
    while (true) {
        // Do normal work 

        if ($runOnlyOnce) {
            break;
        }
    }
}
```

The first solution seems more versatile, while the second one is extremely simple and the exit-logic could also be called before the other already existing exit-checks (like queue empty).

thg2k replied:
I decided to revert the change as the `interrupt()` method inside an event loop callback actually does the job. No need to other another way to achieve the same result.

Namoshek commented on src/MQTTClient.php:
```diff
@@ -58,7 +58,7 @@ class MQTTClient implements ClientContract
 
     /** @var resource|null */
-    private $socket;
+    protected $socket;
```
I'm ok with that but please move it before or after the private members. 👍

thg2k commented:
I decided to revert the change as the `interrupt()` method inside an event loop callback actually does the job. No need to other another way to achieve the same result.

Also moved the socket to the end
